### DISCUSSION TITLE
Help with Lazy Loading

### DISCUSSION THREAD

**Discussion by sapfeer0k** [sourcebroker/t3api#66] [Q&A] [answered]
Hi guys,

Thanks for the cool extension. I tried to start using it, I followed the documentation and create some endpoints. But whenever I try to access them I'm getthing the following exception:
```
Argument 1 passed to SourceBroker\T3api\Service\FileReferenceService::getUrlFromResource() must implement interface TYPO3\CMS\Core\Resource\FileInterface, instance of TYPO3\CMS\Extbase\Persistence\Generic\LazyLoadingProxy given, called in /app/private/typo3conf/ext/t3api/Classes/Serializer/Handler/FileReferenceHandler.php on line 130
```
I'm using LazyLoading for some modules to speed up my app. Any help is appreciated.

**Comment by klamparski** [comment]
Hi @sapfeer0k.

Could you please provide the code of the entity class?

**Reply by sapfeer0k** [reply]
Hi, sure,  here is the link - https://pastebin.com/UJDA0y5y  (password `msALLiBajR`) 
Closely looking, I think it's somehow related to a missing file. but there is no FileReferences defined for this model. 

![image](https://user-images.githubusercontent.com/958172/182389375-ba8dd525-0629-4450-9a85-6a7fd6e451c3.png)

**Reply by klamparski** [reply]
Yes, for sure it's related to file but I would say it's not a missing file. You're right there is no FIleReference in this model but I guess it's in related (nested) model which is passed through serialization process - `Icon`, `Status`, `Module`, `ModuleGroup` or `Account`. Do you have in any of this models FileReference? I would guess that you have somewhere relation to *single* (not collection) file reference with lazy loading applied to it. is it possible? I'm pretty sure it will work fine if lazy is applied to collection but I think I've never applied it to single FileReference. Didn't test it but according to the error message that's the only scenario which I have in mind.

Would be great if you could determine which model causes this error and provide its code again. The easiest way to determine this model would be IMO removing all the properties with relations from `Product` and bring it back one by one and test. According to entity names I guess it will be `Icon`.

**Reply by sapfeer0k** [reply]
Thanks for the help! you were absolutely right. `Module` and `ModuleGroup` are causing the error. 

But what if I don't want to serialize those properties? And I don't need a recursion at all. What's the right wait to configure Normalization / Serialization for only specific fields?  The documentation section is almost empty.
Also, I tried to search into the documentation to switch to `json-ld` format but can't find it.

**Reply by klamparski** [reply, marked as answer]
@sapfeer0k You are right, unfortunately we are missing that part of documentation and I don't really know when we will be able to fill it.
However excluding properties from serialization process is pretty simple and you even already use it for your GET collection endpoint. You just need to configure serialization group for the endpoint and then assign this group to the property you want to include in response.

Endpoint declaration:
```
 *     itemOperations={
 *          "get"={
 *              "path"="/products/{id}",
 *              "normalizationContext"={
 *                  "groups"={"api_product_get_item"}
 *              },
 *          }
 *     },
```

Property declaration:
```
    /**
     * internalName
     *
     * @var string
     * @TYPO3\CMS\Extbase\Annotation\Validate("NotEmpty")
     * @T3api\Serializer\Groups({
     *     "api_product_get_item",
     * })
     */
    protected $internalName = '';
```

When groups are not defined for endpoint then all properties are included.
**I would strongly recommend to always use serialization groups to exclude properties and include them on demand** (by adding group to specific property). It's better for performance and for security reasons because you won't expose accidentally properties. It means: better to define serialization groups for every endpoint but of course it depends on use case.

I've enriched your code by an example (same password as yours). Didn't test it, but I believe it will work.
https://pastebin.com/ANq3wPd0

**Reply by sapfeer0k** [reply]
Thanks again for the clarification! It woks now. 
Could you also suggest how to disable pagination completely and expose `uid` property as `id`?

**Reply by sapfeer0k** [reply]
Also, Could you provide an implementation example of authorization for the API?
